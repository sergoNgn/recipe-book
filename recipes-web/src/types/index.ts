export interface Recipe {
  id?: number;
  name: string;
  shortDescription: string;
  description?: RecipeDescription;
}

export interface RecipeDescription {
  id?: number;
  data: string;
}

export interface Pageable<T> {
  content: T;
  totalElements: number;
  totalPages: number;
}

export enum Validations {
  minLength = "minLength",
  maxLength = "maxLength",
  isEmpty = "isEmpty",
}

export interface Validator {
  validation: Validations;
  errorText: string;
  value?: any;
}

export interface UseInputProps {
  value: string;
  onChange: (event: any) => any;
  onBlur: () => any;
  isDirty: boolean;
  reset: () => any;
  isError: boolean;
  errorText: string;
}

export enum Order {
  ASC = "ASC",
  DESC = "DESC",
}
