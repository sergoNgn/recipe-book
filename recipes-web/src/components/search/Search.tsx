import { FC, useState } from "react";
import { Box, Collapse, TextField, Tooltip } from "@mui/material";
import AddIcon from "@mui/icons-material/AddCircleOutline";
import FilterIcon from "@mui/icons-material/FilterListOutlined";
import UpIcon from "@mui/icons-material/ExpandLess";
import DownIcon from "@mui/icons-material/ExpandMore";
import { Order } from "../../types";

interface SearchProps {
  onAddClick: () => void;
  onSearchValueChange: (value: string) => void;
  onOrderChange: (order: Order) => void;
  order: Order;
}

const Search: FC<SearchProps> = ({
  onAddClick,
  order,
  onSearchValueChange,
  onOrderChange,
}) => {
  const [isFilterCollapsed, setIsFilterCollapsed] = useState(false);
  const [searchValue, setSearchValue] = useState("");
  let timer: any;

  const onFilterClick = () => {
    setIsFilterCollapsed(!isFilterCollapsed);
  };

  const onSearchChange = (event: any) => {
    setSearchValue(event.target.value);
    if (timer) {
      clearTimeout(timer);
    }
    timer = setTimeout(() => onSearchValueChange(event.target.value), 1000);
  };

  const onDescOrder = () => {
    onOrderChange(Order.ASC);
  };

  const onAscOrder = () => {
    onOrderChange(Order.DESC);
  };

  return (
    <Box className="flex flex-row items-center m-2 gap-3">
      <Tooltip placement="top" title="New Recipe">
        <AddIcon className="icon-btn" color="primary" onClick={onAddClick} />
      </Tooltip>
      <Tooltip placement="top" title="Filter">
        <FilterIcon
          className="icon-btn"
          color="primary"
          onClick={onFilterClick}
        />
      </Tooltip>
      <Collapse orientation="horizontal" in={isFilterCollapsed}>
        <TextField
          inputRef={(input) => input && input.focus()}
          size="small"
          variant="outlined"
          className="w-80"
          placeholder="by name"
          autoComplete="off"
          onChange={onSearchChange}
          value={searchValue}
        />
      </Collapse>
      {order === Order.ASC && (
        <Tooltip title="sort by created date asc" placement="top">
          <UpIcon color="primary" className="icon-btn" onClick={onAscOrder} />
        </Tooltip>
      )}
      {order === Order.DESC && (
        <Tooltip title="sort by created date desc" placement="top">
          <DownIcon
            color="primary"
            className="icon-btn"
            onClick={onDescOrder}
          />
        </Tooltip>
      )}
    </Box>
  );
};

export default Search;
