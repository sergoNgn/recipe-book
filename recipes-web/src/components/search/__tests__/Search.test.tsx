import { render } from "@testing-library/react";
import Search from "../Search";
import { Order } from "../../../types";

describe("Search", () => {
  it("renders search component", async () => {
    const alertDialog = render(
      <Search
        onAddClick={() => {}}
        onSearchValueChange={() => {}}
        onOrderChange={() => {}}
        order={Order.ASC}
      />
    );

    expect(alertDialog.baseElement).toMatchSnapshot();
  });
});
