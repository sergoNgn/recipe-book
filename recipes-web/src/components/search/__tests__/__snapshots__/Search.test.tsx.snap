// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders search component 1`] = `
<body>
  <div>
    <div
      class="flex flex-row items-center m-2 gap-3 MuiBox-root css-0"
    >
      <svg
        aria-hidden="true"
        aria-label="New Recipe"
        class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium icon-btn css-ci5apu-MuiSvgIcon-root"
        data-mui-internal-clone-element="true"
        data-testid="AddCircleOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
        />
      </svg>
      <svg
        aria-hidden="true"
        aria-label="Filter"
        class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium icon-btn css-ci5apu-MuiSvgIcon-root"
        data-mui-internal-clone-element="true"
        data-testid="FilterListOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
        />
      </svg>
      <div
        class="MuiCollapse-root MuiCollapse-horizontal MuiCollapse-hidden css-5nave7-MuiCollapse-root"
        style="min-width: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-horizontal css-18w3a48-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-horizontal css-14vyde3-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiFormControl-root MuiTextField-root w-80 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r2:"
                  placeholder="by name"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      â€‹
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
      <svg
        aria-hidden="true"
        aria-label="sort by created date asc"
        class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium icon-btn css-ci5apu-MuiSvgIcon-root"
        data-mui-internal-clone-element="true"
        data-testid="ExpandLessIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
        />
      </svg>
    </div>
  </div>
</body>
`;
